src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -190,42 +208,76 @@ private ReplicationBatch[] PrepareReplicationDataForShards(TransactionOperationC
+                    if (batch.Attachments != null)
+                    {
+                        foreach (var attachment in batch.Attachments)
                         {
-                            Type = ReplicationBatchItem.ReplicationItemType.AttachmentStream,
-                            Base64Hash = attachment.Key,
-                            Stream = new MemoryStream(buffer)
-                        };
-
-                        batches[shard].Attachments[attachment.Key] = attachmentStream;
-
-                        DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Optimization required");
+                            if (attachment.Value == null && _batches.Streams.ContainsKey(attachment.Key))
+                            {
+                                var stream = _batches.Streams[attachment.Key];
+                                AddAttachmentStreamItemToBatch(batch, attachment.Key, stream);
+                            }
+                            else
+                            {
+                                batch.Attachments.Remove(attachment.Key);
+                            }
+                        }
                     }
                 }
             }
 
-            return batches;
+            return _batches;
         }
 
-        private byte[] GetBufferFromAttachmentStream(Stream stream)
+        private void AddAttachmentStreamItemToBatch(ReplicationBatch batch, Slice attachmentKey, Stream stream)
         {
-            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Avoid this");
-
-            var length = (int)stream.Length;
-            var buffer = new byte[length];
+            Stream readerStream;
+            switch (stream)
+            {
+                case MemoryStream ms:
+                    readerStream = new MemoryStream(ms.GetBuffer(), 0, (int)ms.Length);
+                    break;
+                case TempCryptoStream:
+                case TempFileStream:
+                    TempFileStream tmp = stream is TempCryptoStream tcs ? (TempFileStream)tcs.InnerStream : (TempFileStream)stream;
+                    readerStream = tmp.CreateReaderStream();
+                    break;
+                default:
+                    throw new NotSupportedException("Cannot understand how to clone stream: " + stream);
+            }
 
-            stream.Position = 0;
-            int read = stream.Read(buffer, 0, buffer.Length);
-            stream.Position = 0;
+            var attachmentStream = new AttachmentReplicationItem
+            {
+                Type = ReplicationBatchItem.ReplicationItemType.AttachmentStream,
+                Base64Hash = attachmentKey,
+                Stream = readerStream
+            };
 
-            return buffer;
+            attachmentStream.Stream.Seek(0, SeekOrigin.Begin);
+            batch.Attachments[attachmentKey] = attachmentStream;
```

Comment:
let's rename `Attachments` to `AttachmentStreams`

Reply:
Done